src/packages_pull.go

```diff
@@ -0,0 +1,90 @@
+package src
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+)
+
+func PullPackagesForRepo(cacheDir, repoName, ghPatToken, ghcrHost string) error {
+
+	// Get the list of tags for the repo
+	fmt.Fprintf(os.Stdout, "Pulling packages for repo: %s \n", repoName)
+
+	ghPatTokenBase64Encoded := Base64Encode(ghPatToken)
+
+	tags, err := GetPackageTagsListFromGHCR(repoName, ghPatTokenBase64Encoded, ghcrHost)
+	if err != nil {
+		return fmt.Errorf("Error getting list of tags for packages: %s", err)
+	}
+
+	var validTags []string
+	//Pull package for each tag from GHCR
+	for _, tag := range tags {
+		err = PullPackageForTag(repoName, tag, ghPatTokenBase64Encoded, cacheDir, ghcrHost)
+		if err != nil {
+			return fmt.Errorf("Error getting package: %s", err)
+		}
+		validTags = append(validTags, tag)
+	}
+
+	//Write valid package tags in file for pushing to destination
+	err = WriteValidPackageTagsToCache(cacheDir, repoName, validTags)
+	if err != nil {
+		return fmt.Errorf("Error writing valid package tags to file: %s", err)
+	}
+	return nil
+}
+
+func PullPackageForTag(repoName, tagName, ghPatTokenBase64Encoded, cacheDir, ghcrHost string) error {
+
+	// Get the layer digest package
+	targzLayerDigest, err := GetLayerDigestFromGHCR(repoName, tagName, ghPatTokenBase64Encoded, ghcrHost)
+	if err != nil {
+		return fmt.Errorf("Error getting layer digest for tag: %s", err)
+	}
+
+	// Get package from layer digest
+	err = PullPackageFromLayerDigest(repoName, targzLayerDigest, tagName, ghPatTokenBase64Encoded, cacheDir, ghcrHost)
+	if err != nil {
+		return fmt.Errorf("Error pulling package from layer digest: %s", err)
+	}
+
+	return nil
+}
+
+func PullPackageFromLayerDigest(repoName, targzLayerDigestSHA, tagName, ghPatTokenBase64Encoded, cacheDir, ghcrHost string) error {
```

Comment:
I think it is better to move this method to packages_util.
Maintain all the methods which are interacting with ghcr directly at once place

Reply:
The bifurcation of the files are based on the functionality and utility file is not only for GHCR calls.
The pull files (for repository and here for packages) have the functions relevant for pulling whether from dotcom or GHCR.
The utility file has other support functions needed for the pull.
Likewise for push.

This same pattern exists for repo pull/push as well.